### PR TITLE
[`Deberta/Deberta-v2`] Refactor code base to support compile, export, and fix LLM

### LINKED ISSUE #12436: Add DEBERTA-base model for usage in EncoderDecoderModel.
# 🚀 Feature request

Add DEBERTA-base model as an option for creating an EncoderDecoderModel.

## Motivation

Currently only BERT and RoBERTa models can be transformed to a Seq2Seq model via EncoderDecoder class, and for those of use developing DeBERTa models from scratch it would be wonderful to be able to generate a Seq2Seq model from them. Also, the Deberta-base model works much better than BERT and RoBERTa.

### LINKED ISSUE #15216: DebertaForMaskedLM cannot load the parameters in the MLM head from microsoft/deberta-base
my env:

```
transformers    4.15.0
```

```
from transformers import AutoModelForMaskedLM
model = AutoModelForMaskedLM.from_pretrained("microsoft/deberta-base")
```

The warning goes like:

```
Some weights of the model checkpoint at microsoft/deberta-base were not used when initializing DebertaForMaskedLM: ['deberta.embeddings.position_embeddings.weight', 'lm_predictions.lm_head.dense.bias', 'lm_predictions.lm_head.dense.weight', 'lm_predictions.lm_head.bias', 'lm_predictions.lm_head.LayerNorm.weight', 'lm_predictions.lm_head.LayerNorm.bias']
- This IS expected if you are initializing DebertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DebertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DebertaForMaskedLM were not initialized from the model checkpoint at /home/tfangaa/Downloads/ptlm/deberta-base/ and are newly initialized: ['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
```

It seems that the checkpoints provided by `microsoft/deberta-base` doesn't possess the weights needed for the MLM head, so that DebertaForMaskedLM cannot be directly used for masked token prediction.

Is this a bug of the DebertaForMaskedLM class or the checkpoints provided by Microsoft? Thanks!

### LINKED ISSUE #15673: DebertaForMaskedLM cannot load the parameters in the MLM head
Hello,
When I tried to use Deberta for MLM and I didn't succeed. It seems that it is a known issue.  The checkpoint provided by microsoft/deberta-base and other similar ones don't have the pre-trained weights of the Masked LM head.

https://github.com/huggingface/transformers/issues/15216

Thank you!

### LINKED ISSUE #16456: Deberta gives completelly wrong and random output tested on multiple machines and multiple versions of transformers
I do not understand why would you add this model here write lengthy documentation and let it not work at all. 

**Its prediction is: The capital of France is plunge**

I have tried on multiple machines and multiple version of transformers and the results are just random.

    from transformers import pipeline
    unmasker = pipeline('fill-mask', model='deberta-base')
    the_out = unmasker("The capital of France is [MASK].")
    print("the_out",the_out)

As you can see the deberta results is completely wrong, there is some big error in porting it to transformers. 

 the_out [{'score': 0.001861382625065744, 'token': 18929, 'token_str': 'ABC', 'sequence': 'The capital of France isABC.'}, {'score': 0.0012871784856542945, 'token': 15804, 'token_str': ' plunge', 'sequence': 'The capital of France is plunge.'}, {'score': 0.001228992477990687, 'token': 47366, 'token_str': 'amaru', 'sequence': 'The capital of France isamaru.'}, {'score': 0.0010126306442543864, 'token': 46703, 'token_str': 'bians', 'sequence': 'The capital of France isbians.'}, {'score': 0.0008897537481971085, 'token': 43107, 'token_str': 'insured', 'sequence': 'The capital of France isinsured.'}]

    from transformers import pipeline
    unmasker = pipeline('fill-mask', model='bert-base-uncased')
    the_out = unmasker("The capital of France is [MASK].")
    print("the_out",the_out)

The bert result is good:

   the_out [{'score': 0.41678911447525024, 'token': 3000, 'token_str': 'paris', 'sequence': 'the capital of france is paris.'}, {'score':    0.07141649723052979, 'token': 22479, 'token_str': 'lille', 'sequence': 'the capital of france is lille.'}, {'score': 0.06339272856712341, 'token': 10241, 'token_str': 'lyon', 'sequence': 'the capital of france is lyon.'}, {'score': 0.04444753751158714, 'token': 16766, 'token_str': 'marseille', 'sequence': 'the capital of france is marseille.'}, {'score': 0.030297178775072098, 'token': 7562, 'token_str': 'tours', 'sequence': 'the capital of france is tours.'}]

@LysandreJik

### LINKED ISSUE #18659: DeBERTa can't load some parameters
### System Info

- `transformers` version: 4.21.1
- Platform: Linux-5.4.0-81-generic-x86_64-with-glibc2.31
- Python version: 3.9.12
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.11.0+cu113 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@LysandreJik

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

- Reproduction

```python
from transformers import pipeline

text = "The capital of France is [MASK]"

mlm_pipeline = pipeline('fill-mask', model='microsoft/deberta-base', tokenizer='microsoft/deberta-base')
print(mlm_pipeline(text))
```

- Warning Message
```
Some weights of the model checkpoint at microsoft/deberta-base were not used when initializing DebertaForMaskedLM: ['lm_predictions.lm_head.LayerNorm.bias', 'lm_predictions.lm_head.bias', 'lm_predictions.lm_head.dense.bias', 'lm_predictions.lm_head.dense.weight', 'deberta.embeddings.position_embeddings.weight', 'lm_predictions.lm_head.LayerNorm.weight']
- This IS expected if you are initializing DebertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DebertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DebertaForMaskedLM were not initialized from the model checkpoint at microsoft/deberta-base and are newly initialized: ['cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.predictions.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
```

- Output
```
The capital of France isumption
The capital of France is�
The capital of France iszag
The capital of France isreply
The capital of France isnerg
```

### Expected behavior

When DeBERTa model load using transformers, It seems that doesn't load the weights needed for the MLM head. (+ positional embedding weights). 
There are some issues similar to mine.
- https://github.com/huggingface/transformers/issues/15216
- https://github.com/huggingface/transformers/issues/15673
- https://github.com/microsoft/DeBERTa/issues/74
  
But it doesn't seem to be working out yet.  
Can you check it?

### LINKED ISSUE #20815: Cannot export Deberta to TorchScript
### System Info

`transformers-cli env`
```
- `transformers` version: 4.10.2
- Platform: Linux-3.10.0-1127.18.2.el7.x86_64-x86_64-with-glibc2.23
- Python version: 3.9.13
- PyTorch version (GPU?): 1.9.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No
```

### Who can help?

@sgugger

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I am trying to convert the Deberta Model to TorchScript using the instructions provided in the [HF tutorial](https://huggingface.co/docs/transformers/torchscript).

`Code:`
```
import torch
from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("microsoft/deberta-base")
model = AutoModel.from_pretrained("microsoft/deberta-base", torchscript=True)

tokenized_dict = tokenizer(
    ["Is this working",], ["Not yet",], 
    return_tensors="pt"
)
input_tuple = (tokenized_dict['input_ids'], tokenized_dict['attention_mask'])

traced_model = torch.jit.trace(model, input_tuple)
torch.jit.save(traced_model, "compiled_deberta.pt")
```

`Error Message:`
From torch.jit.save:
```
Could not export Python function call 'XSoftmax'. Remove calls to Python functions before export. Did you forget to add @script or @script_method annotation? If this is a nn.ModuleList, add it to __constants__:
```

### Expected behavior

The Traced model should be successfully saved. After loading, it should have the same functional behavior as the model it was traced from.

### LINKED ISSUE #21300: Adding NLLB-200 - MoE - 54.5B for no language left behind
### System Info

Hello  @LysandreJik,
Thanks a lot for your work on no language left behind.

Is there any plan to add the 54.4B Model?

Kindest regards

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Improvement

### Expected behavior

Improvement

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Refactor both Deberta and DebertaV2 to make them more compatible with the overall transformers library

Should fix a bunch of issues related to torch-scripting with Deberta:
fixes #15216
fixes #15673
fixes #16456
fixes #18659
fixes #21300 
fixes #20815
fixes #12436 
fixes #18674

### CODE REVIEW

**Comment by hriaz17** [comment]
Hey @ArthurZucker any updates on this? ETA for when it will be merged into main?

**Comment by ArthurZucker** [comment]
Hey! Just got back from holidays, this should be my main focus in the coming days!

**Comment by ArthurZucker** [comment]
Sorry! Seem like I had to postpone this! If anyone want to take over feel free to do it, otherwise will be my priority once https://github.com/huggingface/transformers/pull/23909 is merge!

**Comment by zynaa** [comment]
Regarding the `z_steps` in `DebertaV2Model`: I think this code is relevant for the [enhanced mask decoder of the generator model](https://github.com/microsoft/DeBERTa/blob/master/DeBERTa/apps/models/masked_language_model.py#L51)

```python
if attention_mask.dim() <= 2:
    extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
    att_mask = extended_attention_mask.byte()
    attention_mask = att_mask * att_mask.squeeze(-2).unsqueeze(-1)
elif attention_mask.dim() == 3:
    attention_mask = attention_mask.unsqueeze(1)
target_mask = target_ids > 0
hidden_states = encoder_layers[-2]
if not self.position_biased_input:
    layers = [encoder.layer[-1] for _ in range(2)]
    z_states += hidden_states
    query_states = z_states
    query_mask = attention_mask
    outputs = []
    rel_embeddings = encoder.get_rel_embedding()

    for layer in layers:
        # TODO: pass relative pos ids
        output = layer(hidden_states, query_mask, return_att=False, query_states=query_states,
                       relative_pos=relative_pos, rel_embeddings=rel_embeddings)
        query_states = output
        outputs.append(query_states)
else:
    outputs = [encoder_layers[-1]]
```

As far as I can tell, they hardcoded z_steps to 2 here. Although it should still be left as 0 for the discriminator. Adding the z_steps to the config seems like a good idea. 

`z_states` represents [the position embeddings](https://github.com/microsoft/DeBERTa/blob/master/DeBERTa/apps/models/masked_language_model.py#L111), which are non-zero if `position_biased_input` is set to `True`. They are passed from the [embedding layer](https://github.com/microsoft/DeBERTa/blob/master/DeBERTa/deberta/bert.py#L269). So in order to properly implement this, I think we need to return the position embeddings here:

```python
class DebertaV2Embeddings(nn.Module):
    def forward(self, input_ids=None, token_type_ids=None, position_ids=None, mask=None, inputs_embeds=None):
        ...

        return embeddings, position_embeddings
```

and implement the `z_steps` like this:

```python
class DebertaV2Model(DebertaV2PreTrainedModel):
    def forward(
        self,
        input_ids: Optional[torch.Tensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        token_type_ids: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.Tensor] = None,
        inputs_embeds: Optional[torch.Tensor] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
    ) -> Union[Tuple, BaseModelOutput]:
        ...

        embedding_output, position_embedding_output = self.embeddings(
            input_ids=input_ids,
            token_type_ids=token_type_ids,
            position_ids=position_ids,
            mask=attention_mask,
            inputs_embeds=inputs_embeds,
        )
        ...

        if self.z_steps > 0:
            hidden_states = encoded_layers[-2]
            layers = [self.encoder.layer[-1] for _ in range(self.z_steps)]
            position_embedding_output += hidden_states
            query_states = position_embedding_output
            query_mask = self.encoder.get_attention_mask(attention_mask)
            rel_embeddings = self.encoder.get_rel_embedding()
            rel_pos = self.encoder.get_rel_pos(embedding_output)
            for layer in layers:
                query_states = layer(
                    hidden_states,
                    query_mask,
                    output_attentions=False,
                    query_states=query_states,
                    relative_pos=rel_pos,
                    rel_embeddings=rel_embeddings,
                )
                encoded_layers = encoded_layers + (query_states,)
```

**Comment by Bachstelze** [comment]
What is the status?
The logs of the checks are expired.

**Comment by ArthurZucker** [comment]
#27734 should help with some of the issues in the mean time

**Comment by joshdevins** [comment]
Any update on this? We're blocked by some of the issues that this is meant to fix. Sepcifically:
* https://github.com/huggingface/transformers/issues/22790
* https://github.com/huggingface/transformers/issues/27586
* https://github.com/huggingface/transformers/issues/20815
* https://github.com/huggingface/transformers/pull/27734

**Comment by serenachou** [comment]
Hi there, just checking in @ArthurZucker on whether there's any progress here please?

**Comment by ArthurZucker** [comment]
Hey hey! Sorry I reviewed the PR but it does not have any activity so I'll just take it over !

**Comment by ArthurZucker** [comment]
I reviewed #27734, will take it over this weekend if possible.

**Comment by ArthurZucker** [comment]
Anyone up for the task can try to do it as well 🤗

**Comment by joshdevins** [comment]
Hey @ArthurZucker any update on the status of this and related issues. Curious if this will also cover DeBERTa v3 now as well.

**Comment by qherreros** [comment]
Hey @ArthurZucker !
Thank you so much for your work. I'm reviewing your changes as it does impact my code on some project I have. I saw that you swap `StableDropout` with `Dropout` in various places. I'm not entirely sure of the impact but it seems to diverge from microsoft implementation. Did you face an issue with `StableDropout` and tracing ? Or is it something else? I'm not super familiar with this part of the code, but if I'll try to help you if I can.

Thanks again

**Comment by ArthurZucker** [comment]
Yep, this decision is based on the fact that `StableDropout` is supposed to be just dropout but more efficient. 
This function is hard to trace / compile and is 4 years old. 
The idea is that if we remove this and XSoftmax, performance gains are so much better and code so much simpler, but equivalent, that is is worth it!

**Comment by qherreros** [comment]
LGTM! Every test we have on our end work with this PR. Thanks @ArthurZucker

**Comment by ArthurZucker** [comment]
cc @michaelbenayoun I am not sure I know how to avoid the HFProxy issue, do you have pointers?

File: src/transformers/models/deberta/configuration_deberta.py [hunk 1]
```diff
@@ -82,6 +82,9 @@ class DebertaConfig(PretrainedConfig):
             `["p2c", "c2p"]`.
         layer_norm_eps (`float`, *optional*, defaults to 1e-12):
             The epsilon used by the layer normalization layers.
+        legacy (`bool`, *optional*, defaults to `True`):
+            Whether or not the model should use the legacy `LegacyDebertaOnlyMLMHead`, which does not work properly
+            for mask infilling tasks.
```
Comment (LysandreJik) [inline src/transformers/models/deberta/configuration_deberta.py:87]: Does this also support the specific dropout/softmax classes or are we dropping them entirely?

**Review by LysandreJik — APPROVED** [review]
Ok! Can you document in DeBERTa's documentation page the evolution that the integration had? I think it's important that users have easily accessible information about the initial contribution, and how this refactor contributes to improving every aspect of DeBERTa.

Thanks!

File: src/transformers/models/deberta/configuration_deberta.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/deberta/configuration_deberta.py:87]: We are dropping them completely!

**Comment by joshdevins** [comment]
Nice one, thanks @ArthurZucker !
